### ISSUE TITLE
Job execution not started with `Iteration is not allowed on @records` error on Dagster 1.8

### ISSUE THREAD

**Issue by mudravrik** [dagster-io/dagster#23698]
### Dagster version

dagster, version 1.8.1

### What's the issue?

TLDR: after updating dagster to 1.8, our executions for assets/jobs fail to start with `Exception: Iteration is not allowed on @records.` error.

We have Dagster Hybrid deployment with dagster libraries fixed in `setup.py` file to version 1.7.
However, it looks like it is either our miss somewhere or just how Dagster Cloud behave - on new deployment we were updated on the side of Dagster Cloud to version 1.8.1.
After that all attempts to start anything fails with following stacktrace:
```
Could not load job definition.
Exception: Iteration is not allowed on `@record`s.
  File "/usr/local/lib/python3.11/site-packages/dagster/_grpc/impl.py", line 135, in core_execute_run
    recon_job.get_definition()
  File "/usr/local/lib/python3.11/site-packages/dagster/_core/definitions/reconstruct.py", line 338, in __hash__
    self._hash = hash_collection(self)
                 ^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 272, in hash_collection
    return hash(make_hashable(collection))
                ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in <listcomp>
    return tuple([make_hashable(x) for x in value])
                  ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in <listcomp>
    return tuple([make_hashable(x) for x in value])
                  ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in <listcomp>
    return tuple([make_hashable(x) for x in value])
                  ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 289, in make_hashable
    return tuple(sorted((key, make_hashable(value)) for key, value in value.items()))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 289, in <genexpr>
    return tuple(sorted((key, make_hashable(value)) for key, value in value.items()))
                              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in <listcomp>
    return tuple([make_hashable(x) for x in value])
                  ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in <listcomp>
    return tuple([make_hashable(x) for x in value])
                  ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 289, in make_hashable
    return tuple(sorted((key, make_hashable(value)) for key, value in value.items()))
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 289, in <genexpr>
    return tuple(sorted((key, make_hashable(value)) for key, value in value.items()))
                              ^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in <listcomp>
    return tuple([make_hashable(x) for x in value])
                  ^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_utils/__init__.py", line 291, in make_hashable
    return tuple([make_hashable(x) for x in value])
                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/dagster/_record/__init__.py", line 443, in _banned_iter
    raise Exception("Iteration is not allowed on `@record`s.")
``` 

I was able to reproduce the error just by rebuild local Docker image with dagster 1.8, however not yet able to locate what exactly causes the error.

So effectively, I have two issues:
- error described above, obviously :)
- is there any way to prevent Dagster Cloud to update its version and respect the one setup in `config.py`?

### Deployment type

Dagster Cloud

### Deployment details

Hybrid deployment with agent on on GCP Compute Engine.

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by gibsondan** [comment]
Hi @mudravrik - we'll investigate the error that you're reporting, but you should be able to control which version of dagster your runs are using. It just runs whatever version you have installed in your Dockerfile, so my suspicion is that something in your image build process is not behaving the way you expect. If you can share your setup.py and Dockerfile we can see if anything jumps out to explain why it's not respecting the pin that you're setting in your setup.py.

**Comment by gibsondan** [comment]
Is there any chance that it's possible to share some definitions code that reproduces the error that you're seeing?

**Comment by mudravrik** [comment]
@gibsondan sure, here it is!

Setup.py:
```
from setuptools import find_packages, setup

setup(
    name="data-platform-dagster",
    packages=find_packages(exclude=["data_platform_dagster_tests", "scripts"]),
    install_requires=[
        "dagit~=1.7",
        "dagster~=1.7",
        "dagster-cloud~=1.7",
        "dagster-dbt",
        "dagster_slack",
        "google-cloud-bigquery[pandas]~=3.11.4",
        "google-cloud-storage~=2.7.0",
        "google-cloud-dataproc~=5.4.1",
        "google-cloud-bigquery-storage",
        "pandas~=1.5",
        "pydantic>=2,<3",
        "python-dateutil~=2.8.2",
        "python-dotenv~=0.21.0",
        "codaio~=0.6.10",
        "requests~=2.31",
        "numpy~=1.24.2",
        "pytz~=2022.7.1",
        "neo4j~=5.6.0",
        "aiohttp~=3.9.2",
        "great_expectations~=0.17",
        "lxml",
        "openai>=1.14,<2",
        "xmltodict",
        "fastavro",
        "tenacity",
    ],
    extras_require={"dev": ["dagster-webserver", "pytest", "flask", "ruff", "pre-commit"]},
)
```

Dockerfile:
```
FROM python:3.11-slim

RUN apt-get update -y && apt-get install -y gcc python3-dev

WORKDIR /opt/dagster/app

COPY . /opt/dagster/app

RUN pip install -e .
```

About defenition, it's a bit more complicated. Commenting out anything but assets (i.e. jobs, sensors and schedules) did not help. Considering assets, I cannot easily take off them one by one without triggering "missing dependency" error. So I cannot yet reproduce the issue on a smaller subset of assets and share it.

If anything unusual, we heavily use dbt-cloud integration and some dynamicaly created assets. Everything else is quite "normal", I think.

**Comment by gibsondan** [comment]
Got it, thanks. Is it possible to run "pip freeze" in the built image and post the output, to double-check that the resulting dagster version is actually 1.8?

**Comment by gibsondan** [comment]
Oh, is it possible that the problem is the `~=`? https://stackoverflow.com/questions/39590187/in-requirements-txt-what-does-tilde-equals-mean suggests to me that that would be the same as

dagster>=1.7, ==1.*

which 1.8 would satsify

**Comment by mudravrik** [comment]
Well. that's quite emberassing. 
I think we fixed dagster version incorrecly:
`~=1.7` is resolved by pip --dry-run to 1.8.1
While we should use `~=1.7.1` to resolve it to the last 1.7.* version.

**Comment by mudravrik** [comment]
> Oh, is it possible that the problem is the `~=`? https://stackoverflow.com/questions/39590187/in-requirements-txt-what-does-tilde-equals-mean suggests to me that that would be the same as
> 
> `dagster>=1.7, ==1.*`
> 
> which `1.8` would satisfy

Yep, that's exactly the issue with Dagster+ version. I'm gonna roll back at least our prod back to 1.7, while looking for root issue cause locally.

**Comment by mudravrik** [comment]
Update: rolled back to 1.7.16, issue does not appears.

**Comment by gibsondan** [comment]
we found what we believe to be the problem here and will be rolling it out in 1.8.2 next week - thanks for the report!

**Comment by mudravrik** [comment]
Thank you so much for such a blazing fast response!
